### ISSUE TITLE
Broken L2ARC space accounting: 16.0E cache device

### ISSUE THREAD

**Issue by jflandry** [zfsonlinux/zfs#3114]
I'm having an issue with some NFS servers, after the cache device fills up the reported size jumps up to 16 Exabytes. If the cache device is removed and re-added the correct size is shown.

```
logs                                                 -      -      -      -      -      -
  scsi-SATA_Crucial_CT240M5_1347095ABE09-part4    140K   160G      0      0    276    106
cache                                                -      -      -      -      -      -
  scsi-SATA_Crucial_CT240M5_1347095ABE0D-part4   2.41T   16.0E      0      9     28   237K
-----------------------------------------------  -----  -----  -----  -----  -----  -----
```

Running on 2.6.32-431.23.3.el6.x86_64 , spl-0.6.3-53_ga3c1eb7 , zfs-0.6.3-163_g9063f65

These have 90 drives hanging off two Supermicro JBODs, 2x Xeon E5-2620, 64GB ram, exporting nfs on IPoIB on QDR Mellanox InfiniBand.

For what it's worth we also have  a couple of OpenVZ hosts with zfs and cache devices on partitions or whole SSDs, only the nfs servers get weird 16E cache devices, but they do get hammered a lot more.

I have found some references from the zfs-discuss and just this morning OmniOS-discuss mailing lists, there seems to be an issue with the upstream ZFS code, the problem is also present on FreeNAS.

Here are the mailing list posts:

https://groups.google.com/a/zfsonlinux.org/forum/#!topic/zfs-discuss/aMwHZrZa5J4/discussion
https://www.mail-archive.com/omnios-discuss@lists.omniti.com/msg03820.html

And a FreeNAS bug report and fix:

https://bugs.freenas.org/issues/6239
https://bugs.freenas.org/projects/freenas/repository/trueos/revisions/6ec48ebf5a1596ec7d2732e891fce3f116105ae5/diff/sys/cddl/contrib/opensolaris/uts/common/fs/zfs/arc.c

One of the NFS servers is for internal use, I could easily test a patch if needed.

**Comment by prakashsurya** [comment]
Hm, it's not obvious why that fixes the problem. Quickly scanning the code, it appears to use the header's "b_asize" field when decrementing the size, so it makes sense to use the same value when incrementing the size accounting. I must be overlooking something..?

**Comment by jflandry** [comment]
@prakashsurya I can only suggest that this may or may not be related to the compressed L2ARC feature.

**Comment by jflandry** [comment]
@prakashsurya Some additional info: I've replaced the cache device on two of the four servers with a full disk SSD and so far everything is behaving normally. I'll keep monitoring for changes.

```
logs                                                -      -      -      -      -      -
  scsi-SATA_Crucial_CT240M5_1328095AA9D6-part4      0   160G      0      0      0      6
cache                                               -      -      -      -      -      -
  scsi-SATA_OCZ-VERTEX2_OCZ-DE9TLH9B1BOUU5VW     223G   793M      0     15  3.36K  1.21M
```

```
logs                                                -      -      -      -      -      -
  scsi-SATA_Crucial_CT240M5_1347095ABDD8-part4   128K   160G      0      0      2      2
cache                                               -      -      -      -      -      -
  scsi-SATA_OCZ-VERTEX2_OCZ-E6K1WP2N49Q19D0U     223G   787M      1     10   157K  1.04M
```

**Comment by kernelOfTruth** [comment]
@jflandry you're still using the l2arc and are affected ?

please give the following commit, patch a try: https://github.com/kernelOfTruth/zfs/commit/d7e1fd0a2b46f38815789ee7c9c425d0801a8d16

**Comment by jflandry** [comment]
@kernelOfTruth That patch seems promising. We've switched the affected servers to whole disk SSDs for now, we've yet to encounter the same issue running with this configuration, but it may just be harder to trigger.

I'm afraid we can't reboot those servers at the moment, except maybe the one for internal use. I'll see if we can patch that one and generate some artificial load.

**Comment by kernelOfTruth** [comment]
*bump*

already somewhat tested patch: https://github.com/zfsonlinux/zfs/pull/3451
(prior to the ARC mutex lock contention fixes merge)

newly updated & "rebased" https://github.com/zfsonlinux/zfs/pull/3491

**Comment by odoucet** [comment]
I've just experienced this bug with very huge L2ARC (two 1.2TB disks). It just took several weeks to fill the whole disks.
Taking offline the devices seems not to work (everything was frozen), so rebot was the only way.
I can try to reproduce it with smaller drives on a second server if it helps.

**Comment by pzwahlen** [comment]
@odoucet From my experience, removing a cache device is a blocking operation (from an IO point of view, at least on zvols). When checked with strace, the 'zpool remove' command blocks on IOCTL "0x5a0c". When my cache was growing to infinity, this could take up to 5 minutes (when my l2 reported size was about 400G). If extrapolating, your 1.2TB cache would need about 15 minutes to be removed.

Now that I run #3491, my cache doesn't grow that much anymore and removing a 60G cache device takes about 15 seconds (but is still a blocking operation).

**Comment by behlendorf** [comment]
@odoucet removing a cache device is a blocking operation.  It doesn't strictly have to be but that's how it was implemented.

Between the following two commits this issue is believe to be resolved in master.

ef56b07 Account for ashift when gathering buffers to be written to l2arc device
d962d5d Illumos 5701 - zpool list reports incorrect "alloc" value for cache device

**Comment by odoucet** [comment]
These two fixes were not merged in 0.6.4.2 ; was it expected ?

**Comment by behlendorf** [comment]
@odoucet they we're deemed too high risk and intentionally skipped.  We want to very conservative in regarding what we backport.
